diff -rupN sqlat-1.1.0/include/sqlanlz.h sqlat-1.1.0-ipv6/include/sqlanlz.h
--- sqlat-1.1.0/include/sqlanlz.h	2004-02-18 12:00:52.000000000 -0500
+++ sqlat-1.1.0-ipv6/include/sqlanlz.h	2013-08-12 10:48:11.000000000 -0400
@@ -21,7 +21,7 @@
 #define _SQLANLZ_H
 
 typedef struct args_struct {
-  char sIP[20];
+  char sIP[50];
   FILE *pIPFile;
   int nPort;
   FILE *pRepFile;
diff -rupN sqlat-1.1.0/include/sqldict.h sqlat-1.1.0-ipv6/include/sqldict.h
--- sqlat-1.1.0/include/sqldict.h	2004-02-18 12:00:52.000000000 -0500
+++ sqlat-1.1.0-ipv6/include/sqldict.h	2013-08-12 10:48:11.000000000 -0400
@@ -20,7 +20,7 @@
 #include <stdio.h>
 
 typedef struct args_struct {
-  char sIP[20];
+  char sIP[50];
   FILE *pUserFile;
   FILE *pPassFile;
   FILE *pIPFile;
diff -rupN sqlat-1.1.0/include/sqldumplogins.h sqlat-1.1.0-ipv6/include/sqldumplogins.h
--- sqlat-1.1.0/include/sqldumplogins.h	2004-02-18 12:00:52.000000000 -0500
+++ sqlat-1.1.0-ipv6/include/sqldumplogins.h	2013-08-12 10:48:11.000000000 -0400
@@ -21,7 +21,7 @@
 #define _SQLDUMPLOGINS_H
 
 typedef struct args_struct {
-  char sIP[20];
+  char sIP[50];
   FILE *pIPFile;
   FILE *pRepFile;
   int  nPort;
diff -rupN sqlat-1.1.0/include/sqlsamdump.h sqlat-1.1.0-ipv6/include/sqlsamdump.h
--- sqlat-1.1.0/include/sqlsamdump.h	2004-02-18 12:00:52.000000000 -0500
+++ sqlat-1.1.0-ipv6/include/sqlsamdump.h	2013-08-12 10:48:11.000000000 -0400
@@ -24,7 +24,7 @@ void tds_free_login(TDSLOGIN *);
 typedef unsigned char byte;
 
 struct args_struct {
-  char sIP[25];
+  char sIP[50];
   FILE *pIPFile;
   int  nPort;
   char sUser[50];
diff -rupN sqlat-1.1.0/include/sqlupload.h sqlat-1.1.0-ipv6/include/sqlupload.h
--- sqlat-1.1.0/include/sqlupload.h	2004-02-18 12:00:52.000000000 -0500
+++ sqlat-1.1.0-ipv6/include/sqlupload.h	2013-08-12 10:48:11.000000000 -0400
@@ -24,7 +24,7 @@ void tds_free_login(TDSLOGIN *);
 typedef unsigned char byte;
 
 struct args_struct {
-  char sIP[20];
+  char sIP[50];
   char filename[1024];
   char sUser[50];
   char sPass[50];
diff -rupN sqlat-1.1.0/sqlanalyze.c sqlat-1.1.0-ipv6/sqlanalyze.c
--- sqlat-1.1.0/sqlanalyze.c	2004-04-20 13:23:41.000000000 -0400
+++ sqlat-1.1.0-ipv6/sqlanalyze.c	2013-08-15 08:21:08.674852157 -0400
@@ -186,12 +186,12 @@ int main(int argc, char **argv) {
   const char* p;
 
   initArguments(&args);
-  if ( ( sIP = malloc(20) ) == NULL ) {
+  if ( ( sIP = malloc(50) ) == NULL ) {
     fprintf(stderr,"sqlanlz: Error allocating sIP buffer\n");
     exit(1);
   }
  
-  memset(sIP,  0, 20);
+  memset(sIP,  0, 50);
   memset(pass, 0, sizeof(pass));
   memset(user, 0, sizeof(user));
 
diff -rupN sqlat-1.1.0/sqldict.c sqlat-1.1.0-ipv6/sqldict.c
--- sqlat-1.1.0/sqldict.c	2004-04-20 13:15:26.000000000 -0400
+++ sqlat-1.1.0-ipv6/sqldict.c	2013-08-12 10:48:11.000000000 -0400
@@ -81,7 +81,7 @@ int doHack(const ARGS *pArgs) {
   TDSCONNECTINFO *connect_info;
   char user[64];
   char pass[64];
-  char ip[16];
+  char ip[50];
   int  exitFlag = FALSE;
   const char* p;
 
@@ -96,7 +96,7 @@ int doHack(const ARGS *pArgs) {
   while (exitFlag == FALSE) {
     
     if (pArgs->pIPFile == NULL) {
-      memset(ip, 0, 16);
+      memset(ip, 0, sizeof(ip));
       strncpy(ip,pArgs->sIP,strlen(pArgs->sIP));
       exitFlag = TRUE;
     } else
diff -rupN sqlat-1.1.0/sqlsamdump.c sqlat-1.1.0-ipv6/sqlsamdump.c
--- sqlat-1.1.0/sqlsamdump.c	2004-04-20 13:15:19.000000000 -0400
+++ sqlat-1.1.0-ipv6/sqlsamdump.c	2013-08-15 08:25:03.834847706 -0400
@@ -297,7 +297,7 @@ int main(int argc, char **argv) {
   int c = 0;
   char *sIP;
 
-  if ((sIP = (char *)malloc(25)) == NULL){
+  if ((sIP = (char *)malloc(50)) == NULL){
     fprintf(stderr,"sqlsamdump: cannot malloc sIP\n");
     exit(1);
   }
